05-Mining-Protocol.md

```diff
@@ -1,101 +1,180 @@
+
+In order to calculate the Merkle Root, an Extended Job carries the following data:
+- `merkle_path`
+- `coinbase_prefix`
+- `coinbase_suffix`
+
+An [Extended Channel](#532-extended-channels) has the following properties:
+- `extranonce_prefix`: the Extended Extranonce bytes that were already allocated by the upstream server.
+- `extranonce_size`: how many bytes are available for the locally reserved and downstream reserved areas of the Extended Extranonce.
+
+And a [Standard Channel](#531-standard-channels) has the following property:
+- `extranonce_prefix` the Extended Extranonce bytes that were already allocated by the upstream server.
+
+So when a proxy receives an Extended Job, it could either:
+- propagate it as an Extended Job to a downstream Extended or Group Channel, where the Extended Extranonce will be further split.
+- convert it into multiple Standard Jobs, where each downstream Standard Channel's `extranonce_prefix` is used to calculate the Merkle Root.
+
+### 5.1.3 Future Jobs
+
+A Job with an empty template or speculated non-empty template can be sent in advance to speedup Job distribution when a new block is found on the network.
+
+The mining server MAY have precomputed such a Job and is able to pre-distribute it for all active Channels.
+The only missing information to start to mine on the new block is the new `prev_hash`.
 This information can be provided independently.
 
 Such an approach improves the efficiency of the protocol where the upstream node does not waste precious time immediately after a new block is found in the network.
 
-To specify that a job is for a future prevhash, the job's `min_ntime` field is left empty.
+The trade-off here is that a non-empty Future Job could potentially contain a transaction that was already included in the block that was just propagated on the network, which would lead to an invalid block if successfully mined.
 
-## 5.2 Hashing Space Distribution
+So mining servers that provide non-empty Future Jobs SHOULD:
+- never send a `SetNewPrevHash` for a Future Job that was later found to contain a conflicting transaction.
+- minimize the probability of Future Jobs containing conflicting transactions (which leads to Future Jobs carrying less profitable templates).
+- after a `SetNewPrevHash` is propagated for a Future Job, provide a non-Future Job for this same `prev_hash` as fast as possible (with a more profitable template).
 
-Each mining device has to work on a unique part of the whole search space.
-The full search space is defined in part by valid values in the following block header fields:
+### 5.1.4 Custom Jobs
 
-- Nonce header field (32 bits)
-- Version header field (16 bits, as specified by [BIP 320](https://github.com/bitcoin/bips/blob/master/bip-0320.mediawiki))
-- Timestamp header field
+Under the Job Declaration Protocol, a Job Declarator Client (JDC) declares a new Job to the Job Declarator Server (JDS), and it ends up with a `mining_job_token`.
```

Comment:
I don't want to get into finer details of the JD protocol here, since this is the spec for the Mining protocol.

For the context of explaining Custom Jobs, it suffices to mention `mining_job_token`. If the reader wants to learn more, they can refer to the JD specs.